### PR TITLE
Manage Topology: Command-line

### PR DESCRIPTION
**Pull request by dokshina**
All commands that allows to manage topology from command-line are implemented.

This is the first part of #396 

I didn't forget about

- [x] Tests 
- [ ] Changelog
- [x] Documentation #412

### CODE REVIEW

File: cli/common/socket.go [hunk 1]
```diff
@@ -91,7 +91,7 @@ func EvalTarantoolConn(conn net.Conn, funcBody string) (interface{}, error) {
 	end
 
 	if err ~= nil then
-		return { success = false, err = err }
+		return { success = false, err = tostring(err) }
```
Comment (olegrok) [inline cli/common/socket.go:94]: Could be moved into separate patch as completely.

File: cli/common/utils.go [hunk 2]
```diff
@@ -238,9 +240,14 @@ Loop:
 	return lastNewLinePos, nil
 }
 
-func ConvertToStringsSlice(s []interface{}) ([]string, error) {
-	stringsSlice := make([]string, len(s))
-	for i, elem := range s {
+func ConvertToStringsSlice(s interface{}) ([]string, error) {
```
Comment (olegrok) [inline cli/common/utils.go:243]: I think a proper name is "ConvertArrayToStringSlice" but I don't insist

File: cli/common/utils.go [hunk 3]
```diff
@@ -274,3 +281,43 @@ func InsertInStringSlice(s []string, i int, elem string) []string {
...
+
+		parts := strings.SplitN(instanceName, ".", 2)
+
+		if len(parts) > 1 {
+			return nil, fmt.Errorf(instanceIDSpecified)
+		}
+
+		if instanceName != "" {
+			if _, found := foundInstances[instanceName]; found {
+				return nil, fmt.Errorf("Duplicate instance name: %s", instanceName)
```
Comment (olegrok) [inline cli/common/utils.go:302]: "... found in config"?

Otherwise it could confuse me as I don't understand the root of problem.

File: cli/replicasets/bootstrap_vshard.go [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+		}
+
+		return err
+	}
+
+	return nil
+}
+
+var (
+	bootstrapVshardBody = `
```
Comment (olegrok) [inline cli/replicasets/bootstrap_vshard.go:52]: Not sure but maybe it's reasonable to call "can_bootstrap_vshard" before (see https://github.com/tarantool/cartridge/blob/e4ecef6fb2176889e4ebcdb99d8a19cbee70c876/cartridge/vshard-utils.lua#L416)

File: cli/replicasets/common.go [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+
+	var allSectionsConf InstancesConf
+	if err := yaml.Unmarshal([]byte(fileContentBytes), &allSectionsConf); err != nil {
+		return nil, fmt.Errorf("Failed to parse instances configuration file %s: %s", ctx.Running.ConfPath, err)
+	}
+
+	instancesConf := make(InstancesConf)
+
+	appInstancePrefix := fmt.Sprintf("%s.", ctx.Project.Name)
+	for key, sectionConf := range allSectionsConf {
```
Comment (olegrok) [inline cli/replicasets/common.go:45]: Should we ignore/consider stateboard here and at https://github.com/tarantool/cartridge-cli/pull/408/files#diff-4fb23dc0ae21422ebe8be6bd721532279e77e346e549a5294f020b4c7443fa62R285 ?

File: cli/replicasets/common.go [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+		}
+	}
+
+	return &instancesConf, nil
+}
+
+func getControlConn(instancesConf *InstancesConf, ctx *context.Ctx, joinInstances []string) (net.Conn, error) {
+	controlInstanceName, err := getJoinedInstanceName(instancesConf, ctx)
+	if err != nil {
+		return nil, fmt.Errorf("Failed find some instance joined to custer")
```
Comment (olegrok) [inline cli/replicasets/common.go:59]: `custer ` -> `cluster `

File: cli/replicasets/common.go [hunk 7]
```diff
@@ -0,0 +1,185 @@
...
+			parts := strings.SplitN(key, ".", 2)
+			instanceName := parts[1]
+			instancesConf[instanceName] = sectionConf
+		}
+	}
+
+	return &instancesConf, nil
+}
+
+func getControlConn(instancesConf *InstancesConf, ctx *context.Ctx, joinInstances []string) (net.Conn, error) {
```
Comment (olegrok) [inline cli/replicasets/common.go:56]: Not sure `getControlConn` should have `joinInstances ` argument. Consider a better name for this function please

File: cli/replicasets/common.go [hunk 8]
```diff
@@ -0,0 +1,185 @@
...
+
+	log.Debugf("Connected to %s", consoleSockPath)
+
+	return conn, nil
+}
+
+func getJoinedInstanceName(instancesConf *InstancesConf, ctx *context.Ctx) (string, error) {
+	joinedInstances, err := getJoinedInstances(instancesConf, ctx)
+	if err != nil {
+		return "", fmt.Errorf("Failed to get instances connected to membership: %s", err)
```
Comment (olegrok) [inline cli/replicasets/common.go:86]: `connected to membership`

I'm not sure it's correct phrase. May be it should be `in membership` (see https://github.com/tarantool/cartridge/blob/dbd7e3385e9afb13f8e754eb562e9d46e60d3217/cartridge/lua-api/get-topology.lua#L156)

File: cli/replicasets/common.go [hunk 9]
```diff
@@ -0,0 +1,185 @@
...
+	}
+
+	var joinedInstanceName string
+	for instanceURI, instance := range *joinedInstances {
+		if instance.UUID != "" {
+			if instance.Alias == "" {
+				return "", fmt.Errorf("Failed to get alias for instance %s", instanceURI)
+			}
```
Comment (olegrok) [inline cli/replicasets/common.go:94]: `if instance.UUID == "" && instance.Alias == ""  { ... }`

or it's valid case when instance doesn't have UUID?

File: cli/replicasets/common.go [hunk 10]
```diff
@@ -0,0 +1,185 @@
...
+	}
+
+	instanceName := runningInstancesNames[0]
+	instanceSocketPath := project.GetInstanceConsoleSock(ctx, runningInstancesNames[0])
+	conn, err := common.ConnectToTarantoolSocket(instanceSocketPath)
+	if err != nil {
+		return nil, fmt.Errorf("Failed to connect to instance %s: %s", instanceName, err)
+	}
+
+	log.Debugf("Connect all replicasets instances to membership")
```
Comment (olegrok) [inline cli/replicasets/common.go:117]: The same as above. Here and below.

File: cli/replicasets/common.go [hunk 11]
```diff
@@ -0,0 +1,185 @@
...
+
+			joinedInstanceName = instance.Alias
+			break
+		}
+	}
+
+	return joinedInstanceName, nil
+}
+
+func getJoinedInstances(instancesConf *InstancesConf, ctx *context.Ctx) (*MembershipInstances, error) {
```
Comment (olegrok) [inline cli/replicasets/common.go:104]: To be honest I don't understand how "join" is related to "membership".

You can found "joined" instances in cartridge topology config. Also seems you need rename "connect to membership" to "probe".

File: cli/replicasets/common.go [hunk 12]
```diff
@@ -0,0 +1,185 @@
...
+}
+
+func convertToSlice(raw interface{}) ([]interface{}, error) {
+	iterfacesSlice, ok := raw.([]interface{})
+	if !ok {
+		return nil, fmt.Errorf("Should be a list, got %#v", raw)
+	}
+
+	return iterfacesSlice, nil
+}
```
Comment (olegrok) [inline cli/replicasets/common.go:185]: Consider moving this code to "utils"

File: cli/replicasets/edit_topology.go [hunk 13]
```diff
@@ -0,0 +1,235 @@
...
+
+type EditInstanceOpts struct {
+	InstanceUUID string
+	Expelled     bool
+}
+
+type EditInstancesOpts []*EditInstanceOpts
+
+func editReplicaset(conn net.Conn, opts *EditReplicasetOpts) (*TopologyReplicaset, error) {
+	// wait for replicaset is healthy
```
Comment (olegrok) [inline cli/replicasets/edit_topology.go:36]: It's not clear how this phrase is related to "editReplicaset" function name

File: cli/replicasets/edit_topology.go [hunk 14]
```diff
@@ -0,0 +1,235 @@
...
+{{ .FormatTopologyReplicasetFunc }}
+
+local res, err = cartridge.admin_edit_topology({
+	replicasets = {
+		{ {{ .ReplicasetInput }} },
+	}
+})
+
+if err ~= nil then
+	return nil, err
```
Comment (olegrok) [inline cli/replicasets/edit_topology.go:204]: This error is an object from "errors" module. Do you consider it?

File: cli/replicasets/edit_topology.go [hunk 15]
```diff
@@ -0,0 +1,235 @@
...
+	joinServerStrings := make([]string, len(instancesURIs))
+	for i, instancesURI := range instancesURIs {
+		joinServerStrings[i] = fmt.Sprintf("{ uri = '%s' }", instancesURI)
+	}
+
+	optString := fmt.Sprintf("%s = { %s }", optName, strings.Join(joinServerStrings, ", "))
+	*optsStrings = append(*optsStrings, optString)
+}
+
+func serializeStringsSlice(stringsSlice []string) string {
```
Comment (olegrok) [inline cli/replicasets/edit_topology.go:182]: I'm sure you need to write general purpose serializer or find ready solution.

Just remember about one place where similar code exists.
https://github.com/tarantool/cartridge-cli/blob/a7387abfe26e5578f97c0bda77d2e120fd57256a/cli/admin/call.go#L136

File: cli/replicasets/join.go [hunk 16]
```diff
@@ -0,0 +1,106 @@
...
+	topologyReplicasets *TopologyReplicasets, instancesConf *InstancesConf) (*EditReplicasetOpts, error) {
+
+	editReplicasetOpts := EditReplicasetOpts{}
+
+	topologyReplicaset := topologyReplicasets.GetByAlias(replicasetAlias)
+	if topologyReplicaset != nil {
+		editReplicasetOpts.ReplicasetUUID = topologyReplicaset.UUID
+	} else {
+		editReplicasetOpts.ReplicasetAlias = replicasetAlias
+	}
```
Comment (olegrok) [inline cli/replicasets/join.go:83]: Why can't you return error here. This case looks strange.

File: cli/replicasets/join.go [hunk 17]
```diff
@@ -0,0 +1,106 @@
...
+
+	return &editReplicasetOpts, nil
+}
+
+func getInstancesURIs(instanceNames []string, instancesConf *InstancesConf) (*[]string, error) {
+	instancesURIs := make([]string, len(instanceNames))
+	for i, instanceName := range instanceNames {
+		instanceConf, found := (*instancesConf)[instanceName]
+		if !found {
+			return nil, fmt.Errorf("Configuration for instance %s isn't found in %s", instanceName, instancesFile)
```
Comment (olegrok) [inline cli/replicasets/join.go:99]: Please check `in` or `at` should be here.

File: cli/replicasets/list.go [hunk 18]
```diff
@@ -0,0 +1,9 @@
+package replicasets
+
+import (
+	"github.com/tarantool/cartridge-cli/cli/context"
+)
+
+func List(ctx *context.Ctx, args []string) error {
+	return nil
```
Comment (olegrok) [inline cli/replicasets/list.go:8]: Where is a code? :)

File: cli/replicasets/membership.go [hunk 19]
```diff
@@ -0,0 +1,134 @@
+package replicasets
```
Comment (olegrok) [inline cli/replicasets/membership.go:1]: Membership is an implementation detail of cartridge. You don't have any guarantees it will work in future or it won't be changed to built-in "swim" module. I'm strictly against this module, I'm strictly against usage of membership. Please, use only cartridge API.

File: cli/replicasets/topology_replicasets.go [hunk 20]
```diff
@@ -0,0 +1,333 @@
...
+
+func getFloatValueFromMap(m map[string]interface{}, key string, valuePtr *float64) error {
+	valueRaw, found := m[key]
+	if !found {
+		return nil
+	}
+
+	var value float64
+
+	switch i := valueRaw.(type) {
```
Comment (olegrok) [inline cli/replicasets/topology_replicasets.go:225]: I think it could be rewrite in more laconic way with reflect.

All int values you can cast with https://golang.org/pkg/reflect/#Value.SetInt and then to float. All float values you can convent with https://golang.org/pkg/reflect/#Value.SetFloat

Also please revise again is it really needed. E.g. json always decode values to float https://golang.org/pkg/encoding/json/#Token (`float64, for JSON numbers`)

File: cli/replicasets/bootstrap_vshard.go (hunk 4 above)
Comment (dokshina) [reply cli/replicasets/bootstrap_vshard.go:52]: I was thinking about it, but `can_bootstrap_vshard` returns just `false` if there is no router or all groups are bootstrapped. But `bootstrap_vshard` returns an error. In fact, `can_bootstrap_vshard` is needed only for front-end to decide if `Bootstrap vshard` button should be drawn or not.

File: cli/replicasets/common.go (hunk 5 above)
Comment (dokshina) [reply cli/replicasets/common.go:45]: Stateboard isn't a part of topology, so I think it should be ignored.
`common. GetInstancesFromArgs` function is also used for starting/stopping. Stateboard isn't passed as an argument, it should be specified by flag.

File: cli/replicasets/weight.go [hunk 21]
```diff
@@ -0,0 +1,72 @@
...
+	}
+
+	instancesConf, err := getInstancesConf(ctx)
+	if err != nil {
+		return fmt.Errorf("Failed to get instances configuration: %s", err)
+	}
+
+	conn, err := getControlConn(instancesConf, ctx, nil)
+	if err != nil {
+		return fmt.Errorf("Failed to connect to Tarantool instance: %s", err)
```
Comment (olegrok) [inline cli/replicasets/weight.go:36]: I'm tired to see such errors. Consider moving connection to tarantool directly to command context. You use it anywhere.

File: cli/replicasets/bootstrap_vshard.go (hunk 4 above)
Comment (olegrok) [reply cli/replicasets/bootstrap_vshard.go:52]: Ok

File: cli/replicasets/common.go (hunk 5 above)
Comment (olegrok) [reply cli/replicasets/common.go:45]: Thanks for explanation

File: cli/common/utils.go (hunk 2 above)
Comment (dokshina) [reply cli/common/utils.go:243]: In fact, it takes interface argument, not array.

File: cli/common/utils.go (hunk 3 above)
Comment (dokshina) [reply cli/common/utils.go:302]: `Duplicate instance name specified`? It's parsing arguments specified by user.

File: cli/replicasets/common.go (hunk 9 above)
Comment (dokshina) [reply cli/replicasets/common.go:94]: This function finds membership instances that have UUIDs (these instances are joined to cluster). 

I think that I need to rework function names and write more comments about this. I will do it.

File: cli/replicasets/edit_topology.go (hunk 15 above)
Comment (dokshina) [reply cli/replicasets/edit_topology.go:182]: Sometimes I will write a separate module for serialization.

File: cli/replicasets/common.go [hunk 22]
```diff
@@ -0,0 +1,185 @@
...
+		if len(joinInstances) > 0 {
+			controlInstanceName = joinInstances[0]
+		}
+	}
+
+	if controlInstanceName == "" {
+		return nil, fmt.Errorf("Failed to find joined instance")
+	}
+
+	consoleSockPath := project.GetInstanceConsoleSock(ctx, controlInstanceName)
```
[the remark this answers is not in the document]
Comment (dokshina) [reply cli/replicasets/common.go:72]: I don't think so. All running instances joined to cluster have available sockets.

File: cli/replicasets/edit_topology.go [hunk 23]
```diff
@@ -0,0 +1,235 @@
...
+	}
+
+	editReplicasetBody, err := templates.GetTemplatedStr(&editReplicasetBodyTemplate, map[string]string{
+		"ReplicasetInput":                  replicasetInput,
+		"FormatTopologyReplicasetFuncName": formatTopologyReplicasetFuncName,
+		"FormatTopologyReplicasetFunc":     formatTopologyReplicasetFunc,
+	})
+
+	if err != nil {
+		return nil, project.InternalError("Failed to get edit topology body by template: %s", err)
```
[the remark this answers is not in the document]
Comment (dokshina) [reply cli/replicasets/edit_topology.go:57]: Ok, I will fix it. But it's internal error. All that you need to understand is that it's my mistake.

File: cli/replicasets/membership.go (hunk 19 above)
Comment (dokshina) [reply cli/replicasets/membership.go:1]: Please, show me the way to do this things without using a membership and I will rewrite this code.

File: cli/repair/clusterwide_config.go [hunk 24]
```diff
@@ -288,7 +288,7 @@ func (topologyConf *TopologyConfType) setReplicasetsConf() error {
 				replicasetConf.LeadersIsString = true
 				replicasetConf.Leaders = append(replicasetConf.Leaders, leadersConverted)
 			case []interface{}:
-				leaders, err := common.ConvertToStringsSlice(leadersConverted)
+				leaders, err := common.ConvertToStringsSlice(leadersRaw)
```
[the remark this answers is not in the document]
Comment (dokshina) [reply cli/repair/clusterwide_config.go:291]: See #411

File: cli/common/socket.go (hunk 1 above)
Comment (dokshina) [reply cli/common/socket.go:94]: See #410

File: cli/replicasets/edit_topology.go (hunk 14 above)
Comment (dokshina) [reply cli/replicasets/edit_topology.go:204]: Yes, `tostring(err)` is called for every error before returning it from `EvalTarantoolConn`.

File: cli/replicasets/join.go (hunk 16 above)
Comment (dokshina) [reply cli/replicasets/join.go:83]: This is very important code (I will add a comment):

* if replicaset already exists, we have to specify it's URI to join new instances to it
* if replicaset doesn't exist, we specify it's alias to create a new replicaset with specified instances

File: cli/common/utils.go (hunk 2 above)
Comment (olegrok) [reply cli/common/utils.go:243]: Yes, but you can convert only arrays

File: cli/common/utils.go (hunk 3 above)
Comment (olegrok) [reply cli/common/utils.go:302]: My bad. Agree

File: cli/replicasets/edit_topology.go (hunk 15 above)
Comment (olegrok) [reply cli/replicasets/edit_topology.go:182]: sometimes means "иногда". But seems you want to say "one day" (однажды).

File: cli/repair/clusterwide_config.go (hunk 24 above)
[the remark this answers is not in the document]
Comment (olegrok) [reply cli/repair/clusterwide_config.go:291]: Great

File: cli/replicasets/common.go (hunk 22 above)
[the remark this answers is not in the document]
Comment (olegrok) [reply cli/replicasets/common.go:72]: Ok

File: cli/replicasets/join.go (hunk 16 above)
Comment (olegrok) [reply cli/replicasets/join.go:83]: Yes, please, do

File: cli/replicasets/membership.go (hunk 19 above)
Comment (olegrok) [reply cli/replicasets/membership.go:1]: Well, it was discussed verbally. Ok.

File: cli/replicasets/weight.go (hunk 21 above)
Comment (dokshina) [reply cli/replicasets/weight.go:36]: I've added `connectToSomeRunningInstance` and `connectToSomeJoinedInstance` functions with comments.

File: cli/replicasets/common.go (hunk 7 above)
Comment (dokshina) [reply cli/replicasets/common.go]: This function is completely removed, see new `common` functions.

File: cli/common/utils.go (hunk 2 above)
Comment (dokshina) [reply cli/common/utils.go]: And function returns error if specified interface isn't array. But in fact it converts interfaces.

File: cli/common/utils.go [hunk 25]
```diff
@@ -257,6 +259,35 @@ func ConvertToStringsSlice(s interface{}) ([]string, error) {
...
+	if !ok {
+		return nil, fmt.Errorf("Should be a map with string keys, got %#v", raw)
+	}
+
+	mapWithStringKeys := make(map[string]interface{})
+
+	for keyRaw, valueRaw := range rawMap {
+		keyString, ok := keyRaw.(string)
+		if !ok {
+			return nil, fmt.Errorf("Has non string key: %#v", keyRaw)
```
Comment (olegrok) [inline cli/common/utils.go:273]: `non string` -> `non-string`

File: cli/replicasets/common.go [hunk 26]
```diff
@@ -0,0 +1,178 @@
...
+			instanceName := parts[1]
+			instancesConf[instanceName] = sectionConf
+		}
+	}
+
+	return &instancesConf, nil
+}
+
+// getJoinedInstanceName returns a name of instancethat is already joined to cluster
+// It gets all membership instances and checks if there is come instance that has
```
Comment (olegrok) [inline cli/replicasets/common.go:102]: `come` -> `some`

File: cli/replicasets/expel.go [hunk 27]
```diff
@@ -0,0 +1,106 @@
...
+	}
+
+	instancesToExpelMap := make(map[string]bool)
+	for _, instanceName := range instancesToExpelNames {
+		instancesToExpelMap[instanceName] = false
+	}
+
+	joinedInstanceName := ""
+
+	// We need to find some  instance that is joined and not specified to expel
```
Comment (olegrok) [inline cli/replicasets/expel.go:39]: `some  instance` -> `some instance`

File: cli/replicasets/replicasets.go [hunk 28]
```diff
@@ -0,0 +1,45 @@
...
+	}
+
+	if ctx.Running.AppDir, err = filepath.Abs(ctx.Running.AppDir); err != nil {
+		return fmt.Errorf("Failed to get application directory absolute path: %s", err)
+	}
+
+	if ctx.Project.Name == "" {
+		if ctx.Project.Name, err = project.DetectName(ctx.Running.AppDir); err != nil {
+			return fmt.Errorf(
+				"Failed to detect application name: %s. Please pass it explicitly via --name ",
```
Comment (olegrok) [inline cli/replicasets/replicasets.go:38]: `via --name ",` -> `via --name",`

File: cli/replicasets/roles.go [hunk 29]
```diff
@@ -0,0 +1,186 @@
...
+	newTopologyReplicaset, err := editReplicaset(conn, editReplicasetOpts)
+	if err != nil {
+		return fmt.Errorf("Failed to update roles list: %s", err)
+	}
+
+	if len(newTopologyReplicaset.Roles) == 0 {
+		log.Infof("Now replicaset %s has no roles enabled", ctx.Replicasets.ReplicasetName)
+	} else {
+		log.Infof(
+			"Replicaset %s now has these roles enabled: ",
```
Comment (olegrok) [inline cli/replicasets/roles.go:146]: `roles enabled: "` -> `roles enabled:"`

File: doc/replicasets.rst [hunk 30]
```diff
@@ -0,0 +1,226 @@
+.. _cartridge-cli.replicasets:
+
+===============================================================================
+Setting up replicasets
+===============================================================================
+
+``cartridge replicasets`` command is used to set up replicasets on local
```
Comment (olegrok) [inline doc/replicasets.rst:7]: `on development` -> `for development`?

File: cli/replicasets/join.go (hunk 17 above)
Comment (dokshina) [reply cli/replicasets/join.go:129]: I've checked - `in`.

File: doc/replicasets.rst (hunk 30 above)
Comment (dokshina) [reply doc/replicasets.rst]: Doc moved to #412 and will be fixed separately.

**Review by olegrok — APPROVED** [review]
